chain/pra/receiver_relay.go

```diff
@@ -0,0 +1,254 @@
+	}
+
+	// For blockproof to work
+	for r.expectMtaHeight < uint64(includeHeader.Number) {
+		stateProofs := make([][]byte, 0)
+
+		gj, bhs, err := r.c.GetJustificationsAndUnknownHeaders(substrate.SubstrateBlockNumber(r.expectMtaHeight + 1))
+		if err != nil {
+			return nil, err
+		}
+
+		r.log.Debugf("buildFinalityProof: found justification at %d by %d", gj.Commit.TargetNumber, r.expectMtaHeight+1)
+
+		blockUpdates, err := r.buildBlockUpdates(uint64(r.expectMtaHeight+1), gj, bhs)
+		if err != nil {
+			return nil, err
+		}
+
+		// Update expectMta for next message
+		r.expectMtaHeight = uint64(gj.Commit.TargetNumber)
+
+		eventGrandpaNewAuthorities, err := r.getGrandpaNewAuthorities(gj.Commit.TargetHash)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(eventGrandpaNewAuthorities) > 0 {
+			r.log.Debugf("buildFinalityProof: found GrandpaNewAuthorities %d", gj.Commit.TargetNumber)
+			newAuthoritiesStateProof, err := r.newStateProof(gj.Commit.TargetHash)
+			if err != nil {
+				return nil, err
+			}
+
+			stateProofs = append(stateProofs, newAuthoritiesStateProof)
+
+			r.expectSetId++
+		}
+
+		r.log.Tracef("newFinalityProofs: lastBlocks %d", r.expectMtaHeight)
+		finalityProof, err := r.newFinalityProof(blockUpdates, stateProofs, []byte{})
+		if err != nil {
+			return nil, err
+		}
+
+		finalityProofs = append(finalityProofs, finalityProof)
+
+		// early exit and only need one StateProof
+		if (r.expectMtaHeight) == uint64(includeHeader.Number) && len(eventGrandpaNewAuthorities) > 0 {
+			return finalityProofs, nil
+		}
+	}
+
+	bp, err := r.newBlockProof(int64(includeHeader.Number), int64(r.expectMtaHeight))
+	if err != nil {
+		return nil, err
+	}
+
+	includedStateProof, err := r.newStateProof(*includeHash)
+	if err != nil {
+		return nil, err
+	}
+
+	stateProofs := [][]byte{includedStateProof}
+
+	finalityProof, err := r.newFinalityProof(nil, stateProofs, bp)
+	if err != nil {
+		return nil, err
+	}
+
+	r.log.Tracef("newFinalityProofs: bp %d", includeHeader.Number)
+	finalityProofs = append(finalityProofs, finalityProof)
+	return finalityProofs, nil
+}
+
+func (r *relayReceiver) newParaFinalityProof(vd *substrate.PersistedValidationData, paraChainId substrate.SubstrateParachainId, paraHead substrate.SubstrateHash, paraHeight uint64) ([][]byte, error) {
+	r.log.Tracef("newParaFinalityProof: paraBlock %d paraHead %s", paraHeight, paraHead.Hex())
+	r.bmvStatus = r.bmvC.GetPraBmvStatus()
+
+	if paraHeight <= uint64(r.bmvStatus.ParaMtaHeight) {
+		return nil, nil
+	}
+
+	includeHeader, includeHash := r.findParasInclusionCandidateIncludedHead(uint64(vd.RelayParentNumber), paraHead, paraChainId)
+
+	if uint64(includeHeader.Number) < r.bmvC.GetRelayMtaOffset() {
+		r.log.Panicf("newParaFinalityProof: includeHeader %d <= relayMtaOffset %d", uint64(includeHeader.Number), r.bmvC.GetRelayMtaOffset())
+	}
+
+	localRelayMtaHeight := r.store.Height()
+
+	// Sync MTA completely
+	// TODO fetch 50 hash per syncing
```

Comment:
@trmaphi Could you please describe in more detail what you want to do?
My ideal:
- The number of concurrent goroutines should not be greater than the number of CPU cores.
- You have to get the block hashes finished before update MTA. (The block hashes loading in concurrent mode must be ordered flowing the block height.)

Reply:
Current implementation:
* Fetch a block hash, then add hash to MTA one by one
Better implementation:
* Fetch multiple block hashes, then add hash to MTA one by one

Reply:
@trmaphi As I have stated. We shouldn't make a pool to retrieve 50 hashes at a time.
For instance, suppose you want to get 1000 block hashes. You should make a pool with a set number of workers (I recommend that the number of workers equals the number of CPU threads). Using 50 workers does not improve performance.

After the workers have gotten all of the block hashes. The block hashes must be sorted by block height. Finally, go through the block hashes, adding each hash to MTA one at a time.

What do you think ?

Reply:
Yes, I'm trying to implement it like that.

And, the substrate client implementation also has the same problem as you pointed out. How could I get suitable number of CPU thread for workerpool number

https://github.com/icon-project/btp/pull/77/files#diff-e37b76c44f291bd86dff768462eb236e64a870a13f4ffed87b6858de2d9087d4R133-R156